near-sdk/src/json_types/hash.rs

```diff
@@ -60,11 +59,45 @@ impl TryFrom<&str> for Base58CryptoHash {
     type Error = Box<dyn std::error::Error>;
 
     fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Ok(value.parse()?)
+    }
+}
+
+impl std::str::FromStr for Base58CryptoHash {
+    type Err = ParseCryptoHashError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
         let mut crypto_hash: CryptoHash = CryptoHash::default();
         let size = bs58::decode(value).into(&mut crypto_hash)?;
         if size != std::mem::size_of::<CryptoHash>() {
-            return Err("Invalid length of the crypto hash (32)".into());
+            return Err(ParseCryptoHashError::InvalidLength(size));
         }
         Ok(Self(crypto_hash))
     }
 }
+
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum ParseCryptoHashError {
+    InvalidLength(usize),
+    Base58(B58Error),
```

Comment:
Note that this makes `B58Error` a public API, so if we chose to switch to a different library to implement this (or even do a major upgrade of Base58), then we'd have to do a major semver bump. I'd personally would try to be concervative with our errors and a) get maximum room for future extensibility b) without exposing internal implementation details. So, something like this:

```rust
pub struct ParseCryptoHashError { kind: ParseCryptoHashErrorKind }

enum ParseCryptoHashErrorKind {
    InvalidLength(usize),
    Base58(B58Error),
}
```

Reply:
yeah probably best to be safe for now, I'll keep the types private for now, and can just expose later if/when solidified